src/FlashLoanVault.sol

```diff
@@ -0,0 +1,285 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.13;
+
+import "lib/yield-utils-v2/contracts/mocks/ERC20Mock.sol";
+import "lib/yield-utils-v2/contracts/token/IERC20.sol";
+import "src/IERC3156FlashLender.sol";
+import "src/IERC3156FlashBorrower.sol";
+import {TransferHelper} from "lib/yield-utils-v2/contracts/token/TransferHelper.sol";
+
+/**
+@title Flash Loan Vault
+@author Calnix
+@dev Contract allows users to exchange a pre-specified ERC20 token for some other wrapped ERC20 tokens.
+@notice Wrapped tokens will be burned, when user withdraws their deposited tokens.
+*/
+
+contract FlashLoanVault is ERC20Mock, IERC3156FlashLender {
+
+    ///@dev Attach library for safetransfer methods
+    using TransferHelper for IERC20;
+
+    ///@dev The keccak256 hash of "ERC3156FlashBorrower.onFlashLoan"
+    bytes32 public constant CALLBACK_SUCCESS = keccak256("ERC3156FlashBorrower.onFlashLoan");
+
+    ///@notice Fee is a constant 0.1%
+    ///@dev 1000 == 0.1%  (1 == 0.0001 %)
+    uint256 public constant fee = 1000; 
+
+    ///@dev mapping of supported addresses by Flash loan provider
+    mapping(address => bool) public supportedTokens;
+
+    ///@dev ERC20 interface specifying token contract functions
+    IERC20 public immutable underlying;    
```

Comment:
You can call this `asset`, which is what ERC4626 intends.
`IERC20 public immutable override asset;`

Is the same in the ABI as:
`function asset() external view returns (address)`

Contract names as types, external vs view, immutable and override are only visible to the compiler. ERCs are about the ABI.

Reply:
Just so I'm clear on what you are saying, help me confirm if this is the correct understanding:

EIP-4262 requires a method `asset` as such: https://github.com/ethereum/EIPs/blob/master/EIPS/eip-4626.md#asset
![image](https://user-images.githubusercontent.com/22549197/171525094-1f63854f-da8e-4f78-bf50-afdb1a756dfb.png)

So instead of implementing both: 
```solidity
    IERC20 public immutable underlying;    
```

```solidity
    ///@notice Function returns contract address of underlying token utilized by the vault (e.g. DAI)
    ///@dev MUST be an ERC-20 token contract 
    function asset() external view returns(address assetTokenAddress){
        assetTokenAddress = address(asset);
    }
```

I should simply implement `IERC20 public immutable override asset;` to achieve the same?

Reply:
That's a good point, which I will raise with the co-writers and EIP editors. As far as I know, there is no need to comply with parameter (input or output) names. If you check ERC20 implementations you will see that there is a large variation as to the parameter names.

Reply:
![image](https://user-images.githubusercontent.com/38806121/171653485-4568d321-e951-44e5-afcc-1853d4ef98a9.png)

Reply:
Thank you for the clarification!